discord/ext/oauth/exchanger.py

```diff
@@ -0,0 +1,120 @@
+
+    Parameters
+    ----------
+    client_id: :class:`int`
+        The client ID of your bot.
+    client_secret: :class:`str`
+        The client secret of your bot.
+    redirect_uri: :class:`str`
+        The redirect URI for your oauth2 application.
+    """
+
+    def __init__(self, client_id: int, client_secret: str, redirect_uri: str) -> None:
+        self.client_id: int = client_id
+        self.client_secret: str = client_secret
+        self.redirect_uri: str = redirect_uri
+
+    async def __aenter__(self) -> Exchanger:
+        headers = {'Content-Type': 'application/x-www-form-urlencoded'}
+        self.__session = aiohttp.ClientSession(headers=headers)
+        self.http_client: HTTPClient = HTTPClient(self.__session.loop, access_token=True)
+        return self
+
+    async def __aexit__(self, *args, **kwargs) -> None:
+        await self.__session.close()
+
+    def expired(self, token: Token, /) -> bool:
+        """Checks if the provided token is expired.
+
+        Parameters
+        ----------
+        token: :class:`Token`
+            The token to check if it is expired.
+        """
+        return token.expired
+
+    @asynccontextmanager
+    async def exchange(self, code: str, /) -> AsyncIterator[Token]:
+        """Exchange a code for an access token.
+
+        Parameters
+        ----------
+        code: :class:`str`
+            The code to exchange.
+        """
+        data: Dict[str, Any] = {
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'grant_type': 'authorization_code',
+            'redirect_uri': self.redirect_uri,
+            'code': code,
+        }
+        async with self.__session.post('https://discordapp.com/api/v9/oauth2/token', data=data) as resp:
+            data = await resp.json()
+        expires_in = data.get('expires_in')
+        if expires_in is not None:
+            data['expires_in'] = datetime.datetime.utcnow() + datetime.timedelta(seconds=expires_in)
+        token = Token(**data)
+        await self.http_client.static_login(str(token))
+        yield token
+
+    @asynccontextmanager
+    async def refresh(self, refresh_token: str, /) -> AsyncIterator[Token]:
+        """Refresh an access token.
+
+        Parameters
+        ----------
+        refresh_token: :class:`str`
+            The refresh token to use.
+        """
+        data: Dict[str, Any] = {
+            'client_id': self.client_id,
+            'client_secret': self.client_secret,
+            'grant_type': 'refresh_token',
+            'refresh_token': refresh_token,
+        }
+        async with self.__session.post('https://discordapp.com/api/v9/oauth2/token', data=data) as resp:
+            data = await resp.json()
+        expires_in = data.get('expires_in')
+        if expires_in is not None:
+            data['expires_in'] = datetime.datetime.utcnow() + datetime.timedelta(seconds=expires_in)
```

Comment:
All library provided datetimes are now [timezone aware](https://docs.python.org/3/library/datetime.html#aware-and-naive-objects) whereas this provides them as naive. This currently goes against the library design.